pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seedmanagement
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenletconfigv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeed represents a Shoot that is registered as Seed.
+type ManagedSeed struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta
+	// Standard list object metadata.
+	metav1.ListMeta
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot is the Shoot that will be registered as Seed.
+	Shoot *Shoot
+	// Seed describes the Seed that will be registered.
+	// Either Seed or Gardenlet must be specified.
+	Seed *SeedTemplateSpec
+	// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+	Gardenlet *Gardenlet
+}
+
+// Shoot identifies the Shoot that will be registered as Seed.
+type Shoot struct {
+	// Name is the name of the Shoot that will be registered as Seed.
+	Name string
+	// APIServer specifies certain kube-apiserver parameters of the Shoot that will be registered as Seed.
+	APIServer *APIServer
```

Comment:
I only added it because I thought maintaining a compatibility with `use-as-seed` is important, and also because these settings might be important for some scenarios. This is also tied to the "back reference" you objected against, it only exists because of this field. I would be happy to remove it (and then the back reference goes away as well), but then a bunch of `botanist` code that adjusts API server replicas based on these settings should be deleted as well. I agree it's a cleaner design, but then it would be clearly incompatible and things might break. Do you have a good idea how to address this?

Reply:
Ok, got it.
Personally, I would prioritize a clean design of the new `seedmanagement` API over compatibility with the annotation.
When we introduce the new API in one of the next releases, we should have a proper release note and documentation about the migration path/incompatibilities between the annotation and the new API.

So my proposal would be to find another suitable replacement for the API server settings of the `use-as-seed` annotation (as you said, they might be important for some use cases), e.g. a new annotation `api-server-settings` or similar which can only be used by shoots, that are referenced by a `ManagedSeed`, to tweak the API server's settings and let's have a clean separation of responsibilities between the Shoot resource/controller and the ManagedSeed resource/controller.
Then we can drop a) the `APIServer` field in the `ManagedSeed` and b) the `ManagedSeedName` field in the `Shoot`.

WDYT?

Reply:
OK, then this is what we should do. Furthermore, as @rfranzke pointed out, the current design has the flaw that if the APIServer settings change in the `ManagedSeed`, simply setting the reference won't trigger a new reconcile since it's already set (this can be fixed by adding a reconcile annotation in this case, but of course this is ugly).

Since it's clear that `APIServer` settings belong to the `Shoot`, not `ManagedSeed`, could we just we put them in the Shoot spec instead of using an annotation, e.g. `spec.apiServer`? I think this is much cleaner and any changes to the spec would trigger a reconcile which is actually needed for the settings to take effect. WDYT?

Reply:
This has the downside that end-users would be able to control these fields as well. We don't want this. The API server shall scale automatically without the necessity/opportunity that the end-user can intervene. 
Yes, we could validate that such settings can only be changed for managed seeds, but it's still ugly to have fields which can only be used in a special case.

I'm even not sure if we still need all of this. We have introduced this 2y back when we didn't have proper autoscaling for the control plane. Now we have HVPA and I would hope/assume that this works properly/good enough for us. So, we could probably also drop it entirely and introduce a "breaking change" here. WDYT?

If you guys want to keep it then I vote for what @timebertt had proposed: Some annotation on the Shoot which is only respected for managed seeds (yes, similarly ugly, but then we can at least remove it in favor of HVPA and other scaling means if we decide that this special replica tweaking for the API servers of managed seeds is no longer needed).

Reply:
> I'm even not sure if we still need all of this. We have introduced this 2y back when we didn't have proper autoscaling for the control plane. Now we have HVPA and I would hope/assume that this works properly/good enough for us. So, we could probably also drop it entirely and introduce a "breaking change" here. WDYT?

Ah, also thought that this might be a legacy thing. If those tweaking options are not as important anymore, then I would also drop them here without a replacement.
@amshuman-kr What is your opinion here? Is HVPA sufficient for our MangedSeeds that we don't need any special tweaking anymore?

Reply:
OK, I will happily remove it then.